internal/pkg/fleetapi/uploader/uploader.go

```diff
@@ -0,0 +1,223 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+// Package uploader contains the methods needed to upload a file using fleet-server's upload endpoints.
+package uploader
+
+import (
+	"bytes"
+	"context"
+	"crypto/sha256"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"math"
+	"net/http"
+	"net/url"
+
+	"github.com/elastic/elastic-agent/internal/pkg/core/backoff"
+	"github.com/elastic/elastic-agent/internal/pkg/core/monitoring/config"
+	"github.com/elastic/elastic-agent/internal/pkg/fleetapi/client"
+)
+
+// The following constants correspond with fleer-server API paths for file uploads.
+const (
+	PathNewUpload    = "/api/fleet/uploads"
+	PathChunk        = "/api/fleet/uploads/%s/%d"
+	PathFinishUpload = "/api/fleet/uploads/%s"
+)
+
+// FileData contains metadata about a file.
+type FileData struct {
+	Size      int64  `json:"size"`
+	Name      string `json:"name"`
+	Extension string `json:"ext"`
+	Mime      string `json:"mime_type"`
+	Hash      struct {
+		SHA256 string `json:"sha256"`
+		MD5    string `json:"md5"`
+	} `json:"hash"`
+}
+
+// NewUploadRequest is the struct that is passed as the request body when starting a new file upload.
+type NewUploadRequest struct {
+	ActionID string     `json:"action_id"`
+	AgentID  string     `json:"agent_id"`
+	Source   string     `json:"src"`
+	File     FileData   `json:"file"`
+	Contents []FileData `json:"contents"`
+}
+
+// NewUploadResponse is the body for the success case when requesting a new file upload.
+type NewUploadResponse struct {
+	UploadID  string `json:"upload_id"`
+	ChunkSize int64  `json:"chunk_size"`
+}
+
+// FinishRequest is the struct that is used when finalizing an upload.
+type FinishRequest struct {
+	TransitHash struct {
+		SHA256 string `json:"sha256"`
+	} `json:"transithash"`
+}
+
+// retrySender wraps the underlying Sender with retry logic.
+type retrySender struct {
+	c    client.Sender
+	max  int
+	wait backoff.Backoff
+}
+
+// Send calls the underlying Sender's Send method.
+// If a non context-related error is returned or the 429 status code is returned the request is retried after a backoff period.
```

Comment:
I don't think any of the upload endpoints return a `429` anymore, however keeping the `retrySender` may be useful in case we want to build rate-limiting into the fleet-server for the future.
Specifically I would like to try to build a streaming version for the upload API so we don't need to pass a buffer to the uploader